hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -4068,8 +4084,31 @@ public void recommissionRegionServer(final ServerName server,
     // Remove the server from decommissioned (draining) server list.
     String parentZnode = getZooKeeper().getZNodePaths().drainingZNode;
     String node = ZNodePaths.joinZNode(parentZnode, server.getServerName());
+    boolean shouldBePermanentlyDecommissioned = false;
+
+    // Get the binary data in the node and check whether the hostname is marked as
+    // permanently decommissioned or not
+    try {
+      byte[] rawData = ZKUtil.getData(getZooKeeper(), node);
+
+      // Check if the data is present for backwards compatibility, some nodes may not have it
+      if (rawData != null && rawData.length > 0) {
+        DrainedZNodeServerData znodeData = DrainedZNodeServerData.parseFrom(rawData);
+        shouldBePermanentlyDecommissioned = znodeData.getMatchHostNameOnly();
+      }
+    } catch (InterruptedException | KeeperException | InvalidProtocolBufferException e) {
+      throw new HBaseIOException(this.zooKeeper.prefix("Unable to recommission "
+        + server.getServerName() + ", was unable to read the node's data"), e);
+    }
+
     try {
-      ZKUtil.deleteNodeFailSilent(getZooKeeper(), node);
+      if (shouldBePermanentlyDecommissioned) {
```

Comment:
There should be a variant of this method that allows the operator to remove a server by hostname only to be removed from the decommissioning list. Unless I'm missing something, once a server is added to the list with this flag enabled, it cannot be removed again.